hir-transform/src/treeify.rs

```diff
@@ -273,63 +276,111 @@ impl RewritePass for Treeify {
+        // treeification.
+        //
+        // The inner loop is responsible for actually treeifying those blocks. This
+        // necessarily has the effect of mutating the function, and therefore requires
+        // us to recompute some of the analyses so that we can properly determine how
+        // to handle certain blocks in the portion of the CFG being treeified, namely
+        // loops (via loop headers).
+        //
+        // Loops require special handling, as during treeification we typically will be
+        // cloning blocks that belong to the portion of the CFG rooted at the block being
+        // treeified. However, if we are treeifying a block that belongs to a loop, we do
+        // not want to clone along control flow edges representing continuation or breaking
+        // out of a loop. On the other hand, if we reach a loop that is strictly dominated
+        // by the block being treeified, we do want to copy those, as each branch of the
+        // tree will need its own copy of that loop.
+        //
+        // To handle this, we require the ability to:
+        //
+        // * Identify loop headers (which requires the loops analysis)
+        // * Identify dominator relationships (is dominated by, does dominate - requires a dominator
+        //   tree)
+        //
+        // The dominator tree requires the control flow graph analysis, and the loop analysis
+        // requires the dominator tree - as a result, each time we modify the CFG, we must also
+        // ensure that all three analyses reflect any effects of such modifications.
+        //
+        // One option: rather than update/recompute all of the analyses, simply keep track
+        // of which original blocks, a newly introduced clone corresponds to. Then, when we
+        // wish to query whether a block is a loop header, or determine a dominator relationship,
+        // we can map blocks back to the original CFG analyses, and answer the questions there.
+        // The only caveat there is to be sure that none of our changes would result in a CFG
+        // where such a mapping would produce incorrect results, but for purposes of treefication
+        // I do not believe that is the case.
+        let mut block_infos = BlockInfos::new(cfg, domtree, loops);
+
+        // For each block B, treeify B IFF it has multiple predecessors, where for each
+        // predecessor P, P dominates B. Treeifying B involves creating a copy of B and
+        // the subgraph of the CFG rooted at B, for each P.
+        //
+        // Loops are handled with a bit more subtlety, in that a loop header may have
+        // multiple predecessors, but only one which dominates the loop header - such
+        // loops are not treeified. However, a loop header with multiple dominating
+        // predecessors _is_ a candidate for treeification, but the loop is only cloned
+        // for each of the dominating predecessors, not all predecessors.
+        //
```

Comment:
I'm not sure I understand this explanation of the difference in handling the loop headers. I mean, looking at the code the only difference I see is a decision to not treeify the loop header when the loop header dominates the root (down below):
```rust
                    if block_infos.is_loop_header(b).is_some()
                        && block_infos.rpo_cmp(b, root).is_lt()
                    {
                        continue;
                    }
```

Reply:
I'll expand on the commentary a bit, a few parts of it are a tad outdated after reviewing them